apps/emqx/src/emqx_log_throttler.erl

```diff
@@ -0,0 +1,138 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_log_throttler).
+
+-behaviour(gen_server).
+
+-include("logger.hrl").
+-include("types.hrl").
+
+-export([start_link/0]).
+
+%% throttler API
+-export([allow/1]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(SEQ_ID(EventId), {?MODULE, EventId}).
+-define(NEW_SEQ, atomics:new(1, [{signed, false}])).
+-define(GET_SEQ(EventId), persistent_term:get(?SEQ_ID(EventId))).
+-define(RESET_SEQ(SeqRef), atomics:put(SeqRef, 1, 0)).
+-define(INC_SEQ(SeqRef), atomics:add(SeqRef, 1, 1)).
+-define(GET_DROPPED(SeqRef), atomics:get(SeqRef, 1) - 1).
+-define(IS_ALLOWED(SeqRef), atomics:add_get(SeqRef, 1, 1) =:= 1).
+
+-define(NEW_THROTTLE(EventId, SeqRef), persistent_term:put(?SEQ_ID(EventId), SeqRef)).
+
+-define(WINDOW_TIME_SEC, emqx:get_config([log, throttling, window_time])).
+
+-spec allow(atom()) -> boolean().
+allow(EventId) when is_atom(EventId) ->
+    Seq = persistent_term:get(?SEQ_ID(EventId), undefined),
+    case Seq of
+        undefined ->
+            %% TODO: this still can flood the logs until throttle seq is created,
+            %% Alternatively we can register all the throttle IDs in init/1,
+            %% but this would require gathering all the throttle Ids somehow...
+            create(EventId),
```

Comment:
This make throttling not very strict: it can pass some log events before the gen_server creates the sequence. 
Though it's probably acceptable I have some ideas how to avoid it completely:
1. Change throttling API, instead of allow/1 implement `log/3,4`, e.g.:
```
log(EventId, Level, Data, Meta) when is_atom(EventId) ->
    Seq = persistent_term:get(?SEQ_ID(EventId), undefined),
    case Seq of
        undefined ->
            %% TODO: this still can flood the logs until throttle seq is created,
            %% Alternatively we can register all the throttle IDs in init/1,
            %% but this would require gathering all the throttle Ids somehow...
            create_and_maybe_log(EventId, Level, Data, Meta),
        SeqRef ->
            maybe_log(?IS_ALLOWED(SeqRef), Level, Data, Meta)
    end.

handle_cast({create_maybe_log, EventId, Level, Data, Meta}, State) ->
    State1 =
        case State of
            #{EventId := _} ->
                %% Must be allready logged: increment counter and silently drop this event   
                ?INC_SEQ(SeqRef),
                State;
            _ ->
                ?SLOG(Level, Data, Meta),
                SeqRef = ?NEW_SEQ,
                %% One event has been already occurred
                ?INC_SEQ(SeqRef),
                ok = ?NEW_THROTTLE(EventId, SeqRef),
                TimerRef = schedule_refresh(?WINDOW_TIME_SEC, EventId),
                State#{EventId => TimerRef}
        end,
    {noreply, State1};
```

2. Create sequences in `init/1`, but then we need to make sure every `event_id` used in ?LOG_THROTTLE somewhere in the code base has its sequence.

Reply:
Maybe not a bad idea to maintain a static list of ids in a hidden config field's default value.
This makes it possible to configure the list.

Reply:
Reworked to use a static list.